File: test_examples/test_example.php
<?php
// PHP test file for funcfinder

namespace App\Example;

// Simple function
function greet(string $name): string {
    return "Hello, $name!";
}

// Function with default parameter
function add(int $a, int $b = 0): int {
    return $a + $b;
}

// Arrow function (PHP 7.4+)
$multiply = fn($a, $b) => $a * $b;

// Class
class Person {
    private string $name;
    private int $age;

    public function __construct(string $name, int $age) {
        $this->name = $name;
        $this->age = $age;
    }

    public function getName(): string {
        return $this->name;
    }

    public function setName(string $name): void {
        $this->name = $name;
    }

    public function describe(): string {
        return "{$this->name}, age {$this->age}";
    }

    public static function create(string $name): Person {
        return new Person($name, 0);
    }

    private function internalMethod(): void {
        // private method
    }
}

// Interface
interface Drawable {
    public function draw(): void;
    public function resize(float $scale): void;
}

// Trait
trait Loggable {
    public function log(string $message): void {
        echo "[LOG] $message\n";
    }

    protected function error(string $message): void {
        echo "[ERROR] $message\n";
    }
}

// Abstract class
abstract class Shape {
    abstract public function area(): float;
    abstract public function perimeter(): float;

    public function describe(): string {
        return "Area: " . $this->area() . ", Perimeter: " . $this->perimeter();
    }
}

// Class using trait and implementing interface
class Circle extends Shape implements Drawable {
    use Loggable;

    private float $radius;

    public function __construct(float $radius) {
        $this->radius = $radius;
    }

    public function area(): float {
        return M_PI * $this->radius * $this->radius;
    }

    public function perimeter(): float {
        return 2 * M_PI * $this->radius;
    }

    public function draw(): void {
        $this->log("Drawing circle with radius {$this->radius}");
    }

    public function resize(float $scale): void {
        $this->radius *= $scale;
    }
}

// Enum (PHP 8.1+)
enum Direction {
    case North;
    case South;
    case East;
    case West;

    public function opposite(): Direction {
        return match($this) {
            self::North => self::South,
            self::South => self::North,
            self::East => self::West,
            self::West => self::East,
        };
    }
}

// Final class
final class Config {
    private static ?Config $instance = null;

    private function __construct() {}

    public static function getInstance(): Config {
        if (self::$instance === null) {
            self::$instance = new Config();
        }
        return self::$instance;
    }

    public function get(string $key): mixed {
        return null;
    }
}

// Anonymous class usage (the anonymous class itself shouldn't be named)
function createLogger(): object {
    return new class {
        public function log(string $msg): void {
            echo $msg;
        }
    };
}

File: test_examples/test_example.rb
# Ruby test file for funcfinder

require 'json'

# Simple function
def greet(name)
  "Hello, #{name}!"
end

# Function with default parameter
def add(a, b = 0)
  a + b
end

# Function with keyword arguments
def create_person(name:, age: 0)
  { name: name, age: age }
end

# Function with block
def with_logging(&block)
  puts "[START]"
  result = block.call
  puts "[END]"
  result
end

# Class
class Person
  attr_accessor :name, :age

  def initialize(name, age = 0)
    @name = name
    @age = age
  end

  def describe
    "#{@name}, age #{@age}"
  end

  def self.create(name)
    new(name)
  end

  private

  def internal_method
    # private method
  end
end

# Module
module Loggable
  def log(message)
    puts "[LOG] #{message}"
  end

  def error(message)
    puts "[ERROR] #{message}"
  end

  module_function

  def module_level_log(message)
    puts "[MODULE] #{message}"
  end
end

# Class with module include
class Logger
  include Loggable

  def initialize(prefix)
    @prefix = prefix
  end

  def info(message)
    log("#{@prefix}: #{message}")
  end
end

# Struct
Point = Struct.new(:x, :y) do
  def distance(other)
    Math.sqrt((x - other.x)**2 + (y - other.y)**2)
  end

  def to_s
    "(#{x}, #{y})"
  end
end

# Class inheritance
class Shape
  def area
    raise NotImplementedError
  end

  def perimeter
    raise NotImplementedError
  end

  def describe
    "Area: #{area}, Perimeter: #{perimeter}"
  end
end

class Circle < Shape
  attr_reader :radius

  def initialize(radius)
    @radius = radius
  end

  def area
    Math::PI * @radius**2
  end

  def perimeter
    2 * Math::PI * @radius
  end
end

class Rectangle < Shape
  attr_reader :width, :height

  def initialize(width, height)
    @width = width
    @height = height
  end

  def area
    @width * @height
  end

  def perimeter
    2 * (@width + @height)
  end
end

# Singleton class
class Config
  @instance = nil

  private_class_method :new

  def self.instance
    @instance ||= new
  end

  def get(key)
    # get config value
  end

  def set(key, value)
    # set config value
  end
end

# Module with class methods
module Calculator
  class << self
    def add(a, b)
      a + b
    end

    def subtract(a, b)
      a - b
    end

    def multiply(a, b)
      a * b
    end
  end
end

# Lambda (should not be detected as function)
multiply = ->(a, b) { a * b }
square = lambda { |x| x * x }

# Proc
double = Proc.new { |x| x * 2 }
